### PR TITLE
ST: fix and new tests for multiple namespace

### PR DESCRIPTION
**Pull request by Frawless**
### Type of change

- Bugfix
- Enhancement / new feature
- Refactoring

### Description

We decided to remove tests for multiple namespaces from **KafkaST** and create new class for these tests. I added changes in CO deployment for watching multiple namespaces according the docs. Based on that I add several new test cases for testing multiple namespaces support.

### Checklist

_Please go through this checklist and make sure all applicable tasks have been done_

- [x] Fix failed tests
- [x] Write tests
- [x] Make sure all tests pass
- [x] Reference relevant issue(s) and close them after merging

### CODE REVIEW

File: api/src/test/java/io/strimzi/test/StrimziExtension.java [hunk 1]
```diff
@@ -606,9 +609,39 @@ protected void after() {
...
+                    Objects.requireNonNull(new File(CO_INSTALL_DIR).listFiles((file, name) -> name.matches("[0-9]*-RoleBinding.*")))
+            ).sorted().collect(Collectors.toMap(file -> file, f -> getContent(f, node -> {
+
+                ArrayNode subjects = (ArrayNode) node.get("subjects");
+                ObjectNode subject = (ObjectNode) subjects.get(0);
+                subject.put("kind", "ServiceAccount")
+                        .put("name", "strimzi-cluster-operator")
+                        .put("namespace", defaultNamespace);
+
+            }), (x, y) -> x, LinkedHashMap::new));
```
Comment (tombentley) [inline api/src/test/java/io/strimzi/test/StrimziExtension.java]: I think we need to stop parsing these resources into `ObjectNode` : It makes it harder to understand what's going on. We should instead parse them into the Fabric8 resource and manipulate them using that typefsafe API.

File: api/src/test/java/io/strimzi/test/StrimziExtension.java [hunk 2]
```diff
@@ -593,6 +593,9 @@ protected void before() {
                     LOGGER.info("Creating namespace '{}' before test per @Namespace annotation on {}", namespace.value(), name(element));
                     kubeClient().createNamespace(namespace.value());
                     previousNamespace = namespace.use() ? kubeClient().namespace(namespace.value()) : kubeClient().namespace();
+                    if (element instanceof Method) {
```
Comment (tombentley) [inline api/src/test/java/io/strimzi/test/StrimziExtension.java]: Is this necessary? I we not support multiple namespaces on classes?

File: systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    void deleteSecondNamespaceResources() {
+        secondNamespaceResources.deleteResources();
+        kubeClient.namespace(DEFAULT_NAMESPACE);
+    }
+
+    private static Resources classResources() {
+        return classResources;
+    }
+
+    private void deployNewTopic() {
```
Comment (tombentley) [inline systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java]: The tests would be eaiser to read if this was explicit about which namespace the topic was created in.

File: systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        kubeClient.namespace(SECOND_NAMESPACE);
+        kubeClient.create(new File(TOPIC_INSTALL_DIR));
+        TestUtils.waitFor("wait for 'my-topic' to be created in Kafka", 120000, 5000, () -> {
+            kubeClient.namespace(DEFAULT_NAMESPACE);
+            List<String> topics2 = listTopicsUsingPodCLI(CLUSTER_NAME, 0);
+            return topics2.contains(TOPIC);
+        });
+    }
+
+    private void deleteNewTopic() {
```
Comment (tombentley) [inline systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java]: Same comment about the namespace this is implicitly using: Include it in the method name, or pass it as an argument.

File: systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        secondNamespaceResources.kafkaEphemeral(kafkaName, 3).done();
+        secondNamespaceResources.kafkaMirrorMaker(CLUSTER_NAME, kafkaSourceName, kafkaTargetName, "my-group", 1, false).done();
+
+        LOGGER.info("Waiting for creation {} in namespace {}", CLUSTER_NAME + "-mirror-maker", SECOND_NAMESPACE);
+        kubeClient.namespace(SECOND_NAMESPACE);
+        kubeClient.waitForDeployment(CLUSTER_NAME + "-mirror-maker", 1);
+    }
+
+
+
```
Comment (tombentley) [inline systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java]: 3 blank lines is 2 too many imho.

File: systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    private static Resources classResources;
+    private static Resources secondNamespaceResources;
+
+
+    /**
+     * Test the case where the TO is configured to watch a different namespace that it is deployed in
+     */
+    @Test
+    @Tag(REGRESSION)
+    void testWatchingOtherNamespace() throws InterruptedException {
```
Comment (tombentley) [inline systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java]: Method name isn't clear about what might be watching multiple NS → `testTopicOperatorWatchingOtherNamespace()`

File: systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+        List<String> topics = listTopicsUsingPodCLI(CLUSTER_NAME, 0);
+        assertThat(topics, not(hasItems(TOPIC)));
+
+        deployNewTopic();
+        deleteNewTopic();
+    }
+
+    /**
+     * Test the case where messages are sent to topic in different namespace
```
Comment (tombentley) [inline systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java]: This doesn't make sense. If you mean a Kafka topic then Kafka has not notion of namespace. If you mean a `KafkaTopic` then it's not clear what you mean by sending messages to it.

File: systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+        checkRecordsForConsumer(MESSAGE_COUNT, jobReadMessagesForTarget);
+        deleteNewTopic();
+    }
+
+    /**
+     * Test the case when Kafka will be deployed in different namespace
+     */
+    @Test
+    @Tag(REGRESSION)
+    void testDeployKafkaToOtherNamespace() throws InterruptedException {
```
Comment (tombentley) [inline systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java]: This is the case where the `Kafka` resource is in a different NS than the CO is running in? Then `testKafkaInDifferentNsThanClusterOperator()` is imho a better name.

File: systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java (hunk 3 above)
Comment (Frawless) [reply systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java]: Namespace and topic name added as an argument.

File: systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java (hunk 4 above)
Comment (Frawless) [reply systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java]: Namespace and topic name added as an argument.

File: systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java (hunk 6 above)
Comment (Frawless) [reply systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java]: Test name changed

File: systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java (hunk 7 above)
Comment (Frawless) [reply systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java]: Test case removed after our discussion.

File: systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java (hunk 8 above)
Comment (Frawless) [reply systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java]: Correct, I renamed that test.

File: systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+    @Test
+    @Tag(REGRESSION)
+    void testKafkaInDifferentNsThanClusterOperator() throws InterruptedException {
+        String kafkaName = kafkaClusterName(CLUSTER_NAME + "-second");
+
+        secondNamespaceResources.kafkaEphemeral(CLUSTER_NAME + "-second", 3).done();
+
+        LOGGER.info("Waiting for creation {} in namespace {}", kafkaName, SECOND_NAMESPACE);
+        kubeClient.namespace(SECOND_NAMESPACE);
+        kubeClient.waitForStatefulSet(kafkaName, 3);
```
Comment (serrss) [inline systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java]: Do we need to set default namespace in `kubeClient` after this line?

File: systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java [hunk 10]
```diff
@@ -0,0 +1,149 @@
...
+        String kafkaName = CLUSTER_NAME + "-target";
+        String kafkaSourceName = kafkaClusterName(CLUSTER_NAME);
+        String kafkaTargetName = kafkaClusterName(kafkaName);
+
+        secondNamespaceResources.kafkaEphemeral(kafkaName, 3).done();
+        secondNamespaceResources.kafkaMirrorMaker(CLUSTER_NAME, kafkaSourceName, kafkaTargetName, "my-group", 1, false).done();
+
+        LOGGER.info("Waiting for creation {} in namespace {}", CLUSTER_NAME + "-mirror-maker", SECOND_NAMESPACE);
+        kubeClient.namespace(SECOND_NAMESPACE);
+        kubeClient.waitForDeployment(CLUSTER_NAME + "-mirror-maker", 1);
```
Comment (serrss) [inline systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java]: The same about default namespace after this line

File: systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java (hunk 9 above)
Comment (serrss) [reply systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java]: Skip my comment, I found that you did it in ` @AfterEach` method

File: systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+    /**
+     * Test the case where the TO is configured to watch a different namespace that it is deployed in
+     */
+    @Test
+    @Tag(REGRESSION)
+    void testTopicOperatorWatchingOtherNamespace() throws InterruptedException {
+        List<String> topics = listTopicsUsingPodCLI(CLUSTER_NAME, 0);
+        assertThat(topics, not(hasItems(TOPIC_NAME)));
+
+        deployNewTopic(SECOND_NAMESPACE, TOPIC_NAME);
```
[the remark this answers is not in the document]
Comment (Frawless) [reply systemtest/src/test/java/io/strimzi/systemtest/MultipleNamespaceST.java]: Actually we don't. In deploy we use wait_for() which cause test failure when the topic is not deployed correctly

**Pull request merged**